proposals/per-extension-language-preferences.md

```diff
@@ -0,0 +1,235 @@
+
+## Motivation
+
+### Objective
+
+Prior to this proposal, browsers automatically selected a `browser.i18n` language for extensions, and developers (users) could not set a specific language.
+
+For multilingual users, sometimes they want to set a specific application (extension) to another language that is different from the system (browser). This need is not uncommon.
+[Android 13](https://developer.android.com/guide/topics/resources/app-languages), [iOS 13 and macOS Catalina](https://developer.apple.com/news/?id=u2cfuj88) support a feature
+called "Per-app language preferences (settings)". This proposal brings the same function to browser extensions.
+
+#### Use Cases
+
+- Extensions create a language selection menu for users.
+- [Optional] Browsers could potentially provide a built-in language selection menu for users, just like Android and iOS.
+- [Optional] Extensions could integrate third-party i18n implementations (frameworks) with the browser's built-in language selection menu.
+
+### Known Consumers
+
+This is a generic i18n feature, which means that all types of extensions can use it. Some extensions already provide this functionality via workarounds or non-browser.i18n implementations.
+
+## Specification
+
+### Definitions
+
+In this document:
+- **locale** or **language code** or **language tag** is a string that represent a language, defined in [BCP 47](https://www.rfc-editor.org/info/bcp47).
+For example, `en-US`, `zh-CN`, `fr`. It is used by `browser.i18n`, `Date`, `Intl` and various other APIs.
+- **the user preferred language** is the language set by `i18n.setCurrentLanguage(code)`. The user may not have set it.
+- **the extension displayed language** is the language which the extension is displayed in. Note that the extension displayed language may be different from the browser UI language. For example, if the user preferred language is not set, the browser UI language is English, and the extension only supports French (default) and Japanese, then the extension displayed language is French because `i18n.getMessage()` returns French message at this situation.
+
+### Schema
+
+```ts
+/**
+ * Get the extension displayed language.
+ * return the language tag, e.g. en-US.
+ */
+i18n.getCurrentLanguage() : string
+
+/**
+ * Set the user preferred language.
+ * if code is null, revert to the unset state.
+ * if code is not valid, return a rejected Promise.
+ * else return a Promise resolved when the operation is complete.
+ */
+i18n.setCurrentLanguage(code: string) : Promise<void>
+
+/**
+ * Get all languages that supported by this extension.
+ * return a Promise resolved with an array of language tags.
+ */
+i18n.getAllLanguages(): Promise<string[]>
+
+/**
+ * After changing to a new language, the browser triggers a language changed event.
+ * The callback is function(code: string) : void.
+ * The code parameter in the callback is the new language.
+ */
+i18n.onLanguageChanged.addListener(callback)
+```
+
+### Behavior
```

Comment:
The `i18n` API is one of the few APIs that are available to all extension contexts, including lower-privileged content scripts.

Could you specify the intended availability of the API?

I'd say, the default is to NOT expose the APIs to content scripts unless a good reason is given otherwise.

The `getCurrentLanguage` method sounds potentially useful to content scripts, but I'm not sure what the value of this would be. It is specified to be synchronous, which means that browser implementations are forced to make the language choice available to all content processes. It is likely for this info to already be available along with the messages though.

`setCurrentLanguage` - To prevent (compromised) content processes from affecting extensions, I would restrict its availability to extension contexts only (e.g. background script, popup, extension tabs, ...).

`getAllLanguages` - given that its primary intended use is to determine the language to change to, I'd also restrict this to non-content scripts too.

`i18n.onLanguageChanged` - sounds potentially useful to content scripts. But even without this, extensions could use extension messaging to notify content scripts upon detecting a language change.

Reply:
For content scripts, there are several possibilities:

1. Loosest: all methods are available in content scripts.
1. Strictest: all methods are not available.
1. Some are not available.

I'm leaning towards option 3. Specifically, `setCurrentLanguage`, `getCurrentLanguage` and `getAllLanguages` are not available in content scripts (at least `setCurrentLanguage` is not available). And `onLanguageChanged` is available.

Reply:
@hanguokai I would suggest `getCurrentLanguage` to be available in content scripts considering the extension might want to use the current language code outside getMessage situations. In addition, if we add some predefined i18n message, something like @@current_locale, this would be available using getMessage in contentScripts anyway.

Reply:
If `onLanguageChanged` is available, I definitely agree with @carlosjeurissen that the corresponding getter should be available too.

Reply:
I add a new section of Content Scripts Availability